### ISSUE TITLE
Clipboard synchronization between wayland and xwayland clients broken

### ISSUE THREAD

**Issue by Ghosthree3** [hyprwm/Hyprland#6132]
### Hyprland Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch main at commit 2ead1fd22103ce065661555513bace5897083ded  (virtual-keyboard: emit event before finishing keyboard).
Date: Sat May 18 05:07:33 2024
Tag: v0.40.0-111-g2ead1fd2, commits: 4717

flags: (if any)


System Information:
System name: Linux
Node name: milkbar
Release: 6.9.1-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Fri, 17 May 2024 16:56:38 +0000


GPU information: 
0b:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Navi 10 [Radeon RX 5600 OEM/5600 XT / 5700/5700 XT] [1002:731f] (rev c1) (prog-if 00 [VGA controller])
0c:00.0 VGA compatible controller [0300]: NVIDIA Corporation TU104 [GeForce RTX 2070 SUPER] [10de:1e84] (rev a1) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:
```

</details>

### Bug or Regression?

Regression

### Description

Trying to copy/paste data between wayland and xwayland clients is broken as of 121d3a7.
```
121d3a72137d4780602cf245704615f63357ea22 is the first bad commit
```

An error can be observed from xwayland clients when trying to paste something copied from a wayland client if the xwayland clipboard is empty.
```
[glfw error 65545]: X11: Failed to convert selection to data from clipboard
```
But if there exists something on the xwayland clipboard (because of a copy from an existing xwayland client) that will be pasted and no error given.
When pasting into a wayland client, the last copy from a wayland client is used (if it still exists), regardless of the existence of a valid xwayland clipboard or not.

There was one time when playing around copying and pasting I managed to crash Hyprland. I think while attempting to close an xwayland window with something on its clipboard, which would hang invisibly.

### How to reproduce

Open two windows, one under wayland, one under xwayland, and try to copy paste between them.

I was testing by opening a terminal (`kitty`) under wayland, then using it to load an x11 copy of itself (`kitty -o linux_display_server=x11`) and trying to copy paste between them.

First noticed because I was unable to copy paste things from my browser into Discord.

**Comment by ding444** [comment]
Can confirm.  I cannot copy/paste between Firefox and electron apps like Discord, Steam, and Obsidian.

<details>
<summary>System info</summary>
```
Hyprland, built from branch main at commit a8522db683ae260f16f8e7778561d8a54906beb1  (keybinds: fix empty on monitor for new workspaces (6089)).
Date: Wed May 15 13:03:51 2024
Tag: v0.40.0-93-ga8522db6, commits: 4699

flags: (if any)
System name: Linux
Node name: endeavour
Release: 6.9.1-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Fri, 17 May 2024 16:56:38 +0000

GPU information:
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation AD102 [GeForce RTX 4090] [10de:2684] (rev a1) (prog-if 00 [VGA controller])

os-release:
NAME="EndeavourOS"
PRETTY_NAME="EndeavourOS"
ID="endeavouros"
ID_LIKE="arch"
BUILD_ID="2024.01.25"
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://endeavouros.com"
DOCUMENTATION_URL="https://discovery.endeavouros.com"
SUPPORT_URL="https://forum.endeavouros.com"
BUG_REPORT_URL="https://forum.endeavouros.com/c/general-system/endeavouros-installation"
PRIVACY_POLICY_URL="https://endeavouros.com/privacy-policy-2"
LOGO="endeavouros"
```
</details>

**Comment by NEUKhoaLe** [comment]
I think Varxy mentioned that Copy/Paste will be broken until the xwayland rewrite is completed.

https://github.com/hyprwm/Hyprland/pull/6086

**Comment by Ghosthree3** [comment]
I suppose this can be closed if he wants in that case, since it's a planned issue.

**Comment by vaxerski** [comment]
we can keep it to avoid dupes, I'll just link 6086 to this

**Comment by The-Briel-Deal** [comment]
Things seem to work for me with XClip? https://github.com/astrand/xclip

Try using this as a workaround for now.

**Comment by RandomLegend** [comment]
@The-Briel-Deal 

Would i have to change the keybindings for copy & paste then aswell? And does XClip work with wayland applications?

**Comment by Ghosthree3** [comment]
I have simply built and am sitting on the commit before the rewrites (4cdddcf) until the xwayland rewrite is merged.

**Comment by kaixenberg** [comment]
Sorry for the dumb question, but how do I clone the repo b4 the commits?

**Comment by Ghosthree3** [comment]
> Sorry for the dumb question, but how do I clone the repo b4 the commits?

```
git clone --recursive https://github.com/hyprwm/Hyprland
cd Hyprland
git checkout 4cdddcf
git submodule update
make all
sudo make install
```
When this is resolved and you want to go back to using your regular installation method (I assume AUR -git), just run `sudo make uninstall` from within the git directory.

**Comment by kaixenberg** [comment]
@Ghosthree3 Thanks. Had figured it out already by searching the web ;-) . Thanks for the idea nd everyone for this grt project

**Comment by jfvillablanca** [comment]
does anyone know how to pin the commit for nix flake inputs? 
i tried this (with submodules=1 based on #5891)
```nix
    hyprland = {
      url = "git+https://github.com/hyprwm/Hyprland/?ref=refs/heads/main&rev=4cdddcfe466cb21db81af0ac39e51cc15f574da9&submodules=1";
    };
```

but it fails
```
error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
error: 4933 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
fatal: clone of 'https://github.com/wolfpld/tracy' into submodule path '/tmp/nix-111478-0/subprojects/tracy' failed
Failed to clone 'subprojects/tracy'. Retry scheduled
error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
error: 1877 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
fatal: clone of 'https://github.com/canihavesomecoffee/udis86' into submodule path '/tmp/nix-111478-0/subprojects/udis86' failed
Failed to clone 'subprojects/udis86'. Retry scheduled
error: 85 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
fatal: clone of 'https://github.com/wolfpld/tracy' into submodule path '/tmp/nix-111478-0/subprojects/tracy' failed
Failed to clone 'subprojects/tracy' a second time, aborting
```

**Comment by navidmafi** [comment]
@jfvillablanca that looks like an unreliable internet connection. git clone is not resilient to connection resets.

>error: RPC failed; curl 92 HTTP/2 stream 5 was not closed cleanly: CANCEL (err 8)
error: 4933 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF

**Comment by andriandreo** [comment]
I'm not precisely an expert here, but just in case this is useful for somebody:

As a researcher, when I need further compatibility with MS Office, I use WPS Office and this issue was bothering me for quite a while. I found the forked script and I modified it a little bit so it's completely functional (I'm not sure why I was getting lots of new line characters when copy-pasting with the original):

 https://gist.github.com/andriandreo/cc48e2d67c389919e18c54429d768c1f

**Comment by Tristan971** [comment]
The issue is closed, but as far as I can tell the issue is still there in 0.41.0.
Should I open a new issue?

**Comment by Agent00Ming** [comment]
there already is an issue open #6247 concerning a specific subset of xwayland applications

**Comment by Nama** [comment]
This is happening to me since v0.41.0 ._.

**Comment by vadim-su** [comment]
Fix doesn't work for me

**Comment by obeobe** [comment]
In my case I needed the clipboard to sync between my VMWare Workstation guest (Ubuntu 24.04, KDE 5, Wayland) and host (Ubuntu 22.04, Gnome, X11). It worked only with some programs. For example, it didn't work when copy-pasting from or to the terminal.

I tried @armenr 's script.

It helped in having text copied on the host sync over to the guest, but not the other way around.

To enable the sync from the guest to the host as well, I amended this:
```
echo -n "$value" | xclip -selection clipboard
```

To this:
```
echo -n "$value" | xclip -selection clipboard
echo -n "$value" | xclip -selection primary
echo -n "$value" | xclip -selection secondary
```

Based on my experimentation only the `secondary` type is needed for syncing to the host, but I added `primary` as well for good measure.

**Comment by alexankitty** [comment]
I ended up tweaking the existing code because it had a habit of copying inconsistently.
I reversed the check to determine the source, and then check if it needs updating.
I also exit early if the inbound is empty to prevent some situations with the clipboard clearing itself.
Then for users with cliphist I added a failsafe in the event that both clipboards appear to be in sync but the clipboard history has not been updated to allow it to update.
I'm sure there are more things that need to be fixed, but the script seems to be stable in my limited testing.

```sh
#!/usr/bin/env sh
#
# Two-way clipboard syncronization between Wayland and X11, with cliphy support!
# !! Recommended use: Drop this file off @ /usr/local/bin/clipsync && make it executable
# Requires: wl-clipboard, xclip, clipnotify.
# Modified from: https://github.com/hyprwm/Hyprland/issues/6132#issuecomment-2127153823
#
# Usage:
#   clipsync watch [with-notifications|without-notifications] - run in background.
#   clipsync stop - kill all background processes.
#   echo -n any | clipsync insert [with-notifications|without-notifications] - insert clipboard content from stdin.
#   clipsync help - display help information.
#
# Workaround for issue:
# "Clipboard synchronization between wayland and xwayland clients broken"
# https://github.com/hyprwm/Hyprland/issues/6132
#
# Also pertains to:
# https://github.com/hyprwm/Hyprland/issues/6247
# https://github.com/hyprwm/Hyprland/issues/6443
# https://github.com/hyprwm/Hyprland/issues/6148

# Updates clipboard content of both Wayland and X11 if current clipboard content differs.
# Usage: echo -e "1\n2" | clipsync insert [with-notifications|without-notifications]
insert() {
  # Read all the piped input into variable.
  value=$(cat)
  wValue=$(wl-paste -n 2>/dev/null || echo "")
  xValue=$(xclip -o -selection clipboard 2>/dev/null || echo "")

  notify() {
    if [ "$1" != "without-notifications" ]; then
      notify-send -u low -c clipboard "$2" "$value"
    fi
  }

  # discard if empty
  if [ -z "$value" ]; then
    exit
  fi

  if command -v cliphist >/dev/null 2>&1; then
    clipHistLast=$(cliphist list | head -1 | cliphist decode)
    if [ "$value" != "$clipHistLast" ]; then
      notify "$1" "Failsafe Copy"
      echo -n "$value" | wl-copy
      echo -n "$value" | cliphist store
    fi
  fi


  if [ "$value" == "$wValue" ]; then
    if [ "$value" != "$xValue" ]; then
      notify "$1" "Wayland"
      echo -n "$value" | xclip -selection clipboard
      # Add to cliphist if it's installed
      if command -v cliphist >/dev/null 2>&1; then
        echo -n "$value" | cliphist store
      fi
      exit
    fi
  fi

  if [ "$value" == "$xValue" ]; then
    if [ "$value" != "$wValue" ]; then
      notify "$1" "X11"
      echo -n "$value" | wl-copy
      # Add to cliphist if it's installed
      if command -v cliphist >/dev/null 2>&1; then
        echo -n "$value" | cliphist store
      fi
      exit
    fi
  fi
}

# Watch for clipboard changes and synchronize between Wayland and X11
# Usage: clipsync watch [with-notifications|without-notifications]
watch() {
  # Add a small delay to ensure clipboard services are initialized
  sleep 1

  notification_mode=${1:-with-notifications}

  # Wayland -> X11
  wl-paste --type text --watch bash -c "clipsync insert $notification_mode" &

  # X11 -> Wayland
  while clipnotify; do
    #notify-send "clip" $(xclip -o -selection clipboard 2>/dev/null)
    #notify-send "primary" $(xclip -o -selection primary 2>/dev/null)
    #notify-send "secondary " $(xclip -o -selection secondary 2>/dev/null)

    xclip -o -selection clipboard 2>/dev/null | clipsync insert "$notification_mode"
  done &
}

# Kill all background processes related to clipsync
stop_clipsync() {
  pkill -f "wl-paste --type text --watch"
  pkill clipnotify
  pkill -f "xclip -selection clipboard"
  pkill -f "clipsync insert"
}

help() {
  cat << EOF
clipsync - Two-way clipboard synchronization between Wayland and X11, with cliphist support

Usage:
  clipsync watch [with-notifications|without-notifications]
    Run clipboard synchronization in the background.
    Options:
      with-notifications (default): Show desktop notifications for clipboard changes.
      without-notifications: Operate silently without notifications.

  clipsync stop
    Stop all background processes related to clipsync.

  echo -n "text" | clipsync insert [with-notifications|without-notifications]
    Insert clipboard content from stdin.
    Notification options work the same as in the watch command.

  clipsync help
    Display this help information.

Requirements: wl-clipboard, xclip, clipnotify
Optional: cliphist (for Hyprland users)
EOF
}

case "$1" in
  watch)
    watch "$2"
    ;;
  stop)
    stop_clipsync
    ;;
  insert)
    insert "$2"
    ;;
  help)
    help
    ;;
  *)
    echo "Usage: $0 {watch [with-notifications|without-notifications]|stop|insert [with-notifications|without-notifications]|help}"
    echo "Run '$0 help' for more information."
    exit 1
    ;;
esac
```

**Comment by codevski** [comment]
not sure why this was closed seems like people still having issues...

**Comment by arabianq** [comment]
None of the solutions suited me completely, so I wrote my synchronization in python.
It requires xclip, wl-clipboard and clipnotify to work.

https://github.com/arabianq/wl-x11-clipsync

**Comment by JunYang-tes** [comment]
> None of the solutions suited me completely, so I wrote my own synchronization in python. It requires xclip, wl-clipboard and clipnotify to work.
> 
> https://github.com/arabianq/wl-x11-clipsync

prefect!  please note this require python version bigger than 3.9

**Comment by eisterman** [comment]
Is there still not an official solution from Hyprland? :c

**Comment by worksasintended** [comment]
Issue still seems to persist, making it very hard to use in a productive environment...

**Comment by vadim-su** [comment]
> I can confirm this issue has been fixed in 0.49.0

Cool. I will check it today.

**Comment by 24bit192kHz** [comment]
> I can confirm this issue has been fixed in 0.49.0

no it's not
```
❯ yay -Q brave
brave-bin 1:1.78.97-1
❯ hyprland -v
Hyprland 0.49.0 built from branch  at commit 9958d297641b5c84dcff93f9039d80a5ad37ab00  (version: bump to 0.49.0).
Date: Thu May 8 20:15:18 2025
Tag: v0.49.0, commits: 6098
built against:
 aquamarine 0.8.0
 hyprlang 0.6.3
 hyprutils 0.7.1
 hyprcursor 0.1.12
 hyprgraphics 0.1.3


no flags were set
```

**Comment by joelmathewthomas** [comment]
> > I can confirm this issue has been fixed in 0.49.0
> 
> no it's not
> 
> ```
> ❯ yay -Q brave
> brave-bin 1:1.78.97-1
> ❯ hyprland -v
> Hyprland 0.49.0 built from branch  at commit 9958d297641b5c84dcff93f9039d80a5ad37ab00  (version: bump to 0.49.0).
> Date: Thu May 8 20:15:18 2025
> Tag: v0.49.0, commits: 6098
> built against:
>  aquamarine 0.8.0
>  hyprlang 0.6.3
>  hyprutils 0.7.1
>  hyprcursor 0.1.12
>  hyprgraphics 0.1.3
> 
> 
> no flags were set
> ```

How can i repo?

**Comment by c4nc** [comment]
I'm still experiencing this issue, any updates? 

**ENV**
Arch linux

[Brave 1.80.108 (Official Build) beta (64-bit)]
Chromium: 137.0.7151.104

Wayland: 1.23.1-2
xorg-xwayland 24.1.7-1

Copy text from any wayland app, moving focus on xwayland app make the system freeze and then crash.

**Comment by ibrokemypie** [comment]
Whats the current best solution for this? Currently running https://github.com/arabianq/wl-x11-clipsync however its pretty unreliable and I frequently find items I copied don't make it to the clipboard, and/or errors in my journal like this:

```
Jun 20 13:55:47 noodlepad clipsync[4633]: Traceback (most recent call last):
Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/cl60273r4mnq9ax77m4dyzzd1h03xhdk-wl-x11-clipsync-fc3ac4d1d57ffdc3222e818c8a58d20c91f3fcf3/bin/.clipsync-wrapped", line 87, in get_wayland_clipboard
Jun 20 13:55:47 noodlepad clipsync[4633]:     out = subprocess.run(['wl-paste', '-t', mime],
Jun 20 13:55:47 noodlepad clipsync[4633]:           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/lib/python3.12/subprocess.py", line 550, in run
Jun 20 13:55:47 noodlepad clipsync[4633]:     stdout, stderr = process.communicate(input, timeout=timeout)
Jun 20 13:55:47 noodlepad clipsync[4633]:                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/lib/python3.12/subprocess.py", line 1209, in communicate
Jun 20 13:55:47 noodlepad clipsync[4633]:     stdout, stderr = self._communicate(input, endtime, timeout)
Jun 20 13:55:47 noodlepad clipsync[4633]:                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/lib/python3.12/subprocess.py", line 2116, in _communicate
Jun 20 13:55:47 noodlepad clipsync[4633]:     self._check_timeout(endtime, orig_timeout, stdout, stderr)
Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/lib/python3.12/subprocess.py", line 1253, in _check_timeout
Jun 20 13:55:47 noodlepad clipsync[4633]:     raise TimeoutExpired(
Jun 20 13:55:47 noodlepad clipsync[4633]: subprocess.TimeoutExpired: Command '['wl-paste', '-t', 'text/plain']' timed out after 0.8 seconds
```

**Comment by BigPapaAnimal** [comment]
> I'm still experiencing this issue, any updates?
> 
> **ENV** Arch linux
> 
> [Brave 1.80.108 (Official Build) beta (64-bit)] Chromium: 137.0.7151.104
> 
> Wayland: 1.23.1-2 xorg-xwayland 24.1.7-1
> 
> Copy text from any wayland app, moving focus on xwayland app make the system freeze and then crash.

I just started getting this error two days ago and I just cant seem to find a solution, really frustrating.

**Comment by Diamantino-Op** [comment]
Same here, i still get the error

**Comment by ibrokemypie** [comment]
> Whats the current best solution for this? Currently running https://github.com/arabianq/wl-x11-clipsync however its pretty unreliable and I frequently find items I copied don't make it to the clipboard, and/or errors in my journal like this:
> 
> ```
> Jun 20 13:55:47 noodlepad clipsync[4633]: Traceback (most recent call last):
> Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/cl60273r4mnq9ax77m4dyzzd1h03xhdk-wl-x11-clipsync-fc3ac4d1d57ffdc3222e818c8a58d20c91f3fcf3/bin/.clipsync-wrapped", line 87, in get_wayland_clipboard
> Jun 20 13:55:47 noodlepad clipsync[4633]:     out = subprocess.run(['wl-paste', '-t', mime],
> Jun 20 13:55:47 noodlepad clipsync[4633]:           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/lib/python3.12/subprocess.py", line 550, in run
> Jun 20 13:55:47 noodlepad clipsync[4633]:     stdout, stderr = process.communicate(input, timeout=timeout)
> Jun 20 13:55:47 noodlepad clipsync[4633]:                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/lib/python3.12/subprocess.py", line 1209, in communicate
> Jun 20 13:55:47 noodlepad clipsync[4633]:     stdout, stderr = self._communicate(input, endtime, timeout)
> Jun 20 13:55:47 noodlepad clipsync[4633]:                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/lib/python3.12/subprocess.py", line 2116, in _communicate
> Jun 20 13:55:47 noodlepad clipsync[4633]:     self._check_timeout(endtime, orig_timeout, stdout, stderr)
> Jun 20 13:55:47 noodlepad clipsync[4633]:   File "/nix/store/8w718rm43x7z73xhw9d6vh8s4snrq67h-python3-3.12.10/lib/python3.12/subprocess.py", line 1253, in _check_timeout
> Jun 20 13:55:47 noodlepad clipsync[4633]:     raise TimeoutExpired(
> Jun 20 13:55:47 noodlepad clipsync[4633]: subprocess.TimeoutExpired: Command '['wl-paste', '-t', 'text/plain']' timed out after 0.8 seconds
> ```

Opened an issue on the repo for this: https://github.com/arabianq/wl-x11-clipsync/issues/1

After some time my clipboard just freaks out and trying to paste will usually just freeze the paste target app for a few seconds rather than actually pasting anything, very frustrating! All ears if anyone has a solution.

**Comment by MiloLug** [comment]
Same here. After an hour or so of uptime, the clipboard stops working, but it's actually a one-sided issue for me.
For example, I can copy from xwayland apps to wayland, but not the other way around. Also it continues to work between x apps themselves.
What's interesting, clipse stops saving new entries at all

**Comment by baykalokandemir** [comment]
I suffer from the same problem, but on top of not being able to copy from xwayland windows, attempting to copy and paste from an xwayland seems to stop cliphist from working. It won't save any new items copied (from a wayland window) after attempting to copy something from an xwayland window. This could be a cliphist bug.

**Comment by Diamantino-Op** [comment]
> Same here. After an hour or so of uptime, the clipboard stops working, but it's actually a one-sided issue for me. For example, I can copy from xwayland apps to wayland, but not the other way around. Also it continues to work between x apps themselves. What's interesting, clipse stops saving new entries at all

yeah, mine does that too, i can copy from xwayland to wayland, but not the other way around

**Comment by OliverKreitmeyer** [comment]
Still experiencing this issue on 6.16.4-arch1-1 with Hyprland 0.50.1 built from branch  at commit `4e242d086e20b32951fdc0ebcbfb4d41b5be8dcc`.

Spent a good part of 2 days trying to figure it out with no luck. Is the solve for this in the works or should I try and find a workaround?

**Comment by olivatooo** [comment]
> Wayland is the future
> looks inside
> clipboard does not work

**Comment by Ghosthree3** [comment]
This issue is long since fixed, I've not had any problem with the clipboard sync between wayland and xwayland for many months now. I do not use any form of clipboard manager.

I've just tested it using the sample config for Hyprland directly (`Hyprland -c /usr/share/hypr/hyprland.conf`), I did not start any programs or services beyond Kitty, Firefox, and Discord. I was able to copy text between Firefox (wayland) and Discord (xwayland) in both directions without any problem.

If you are still having problems, it's likely being caused by something else you're running, I'd advise you try what I did to see if the problem is actually Hyprland or not.

**Comment by OliverKreitmeyer** [comment]
I'm not sure... I have tried running Hyprland by rebooting, immediately pressing ctrl+alt+F3 to enter the command-line and log in without a login manager, started Hyprland with `Hyprland -c /usr/share/hypr/hyprland.conf` and opened nothing except kitty, with which I launched Obsidian on Wayland, and Thorium (chromium fork) on xwayland. 
I copied a jpeg from Thorium into Obsidian, and the issue persists.

**Comment by Ghosthree3** [comment]
What about regular text?

**Comment by OliverKreitmeyer** [comment]
Regular text works just fine. If it's not complete clipboard synchronization, it might be something to do with image parsing?
I've noticed .png images work fine, however .jpeg crashes the app...

I will keep on troubleshooting this tomorrow, and post if I find out anything.

**Comment by Xenomorphy** [comment]
> Regular text works just fine. If it's not complete clipboard synchronization, it might be something to do with image parsing? I've noticed .png images work fine, however .jpeg crashes the app...
> 
> I will keep on troubleshooting this tomorrow, and post if I find out anything.

Almsot the same situation, jpeg's fail. Text and png works fine. It doesn't crash the app for me but it just don't paste anything at all.

**Comment by j0ntsao** [comment]
Still having issues pasting pngs from xwayland apps to wayland, e.g. large pngs copied from spyder to firefox or slack are an issue but pngs from other wayland apps are fine.
